qiskit/transpiler/passes/routing/basic_swap.py

```diff
@@ -98,12 +108,79 @@ def run(self, dag):
                     # update current_layout
                     for swap in range(len(path) - 2):
                         current_layout.swap(path[swap], path[swap + 1])
-
-            order = current_layout.reorder_bits(new_dag.qubits)
-            new_dag.compose(subdag, qubits=order)
+            # handle control flow operations
+            for node in subdag.control_flow_ops():
+                updated_ctrl_op, cf_layout = self._transpile_controlflow_op(node.op, current_layout)
+                node.op = updated_ctrl_op
+                cf_layer = True
+            if cf_layer:
+                new_dag.compose(subdag)
+                current_layout = cf_layout
+            else:
+                order = current_layout.reorder_bits(new_dag.qubits)
+                new_dag.compose(subdag, qubits=order)
+        self.property_set["final_layout"] = current_layout
 
         return new_dag
 
+    def _transpile_controlflow_op(self, cf_op, current_layout):
+        """default function"""
+        if isinstance(cf_op, IfElseOp):
+            return self._transpile_controlflow_multiblock(cf_op, current_layout)
+        elif isinstance(cf_op, (ForLoopOp, WhileLoopOp)):
+            return self._transpile_controlflow_looping(cf_op, current_layout)
+        return cf_op, current_layout
+
+    def _transpile_controlflow_multiblock(self, cf_op, current_layout):
+        # pylint: disable=cyclic-import
+        from qiskit.transpiler.passes.routing.layout_transformation import LayoutTransformation
+        from qiskit.converters import dag_to_circuit, circuit_to_dag
+
+        block_circuits = []  # control flow circuit blocks
+        block_dags = []  # control flow dag blocks
+        block_layouts = []  # control flow layouts
+
+        for i, block in enumerate(cf_op.blocks):
+            dag_block = circuit_to_dag(block)
+            _pass = BasicSwap(self.coupling_map, initial_layout=current_layout)
+            updated_dag_block = _pass.run(dag_block)
+            block_dags.append(updated_dag_block)
```

Comment:
Do you know what will happen if `BasisSwap` inserts a `SwapGate` on a qubit not in `cf_op.qargs`?

Reply:
`cf_op.qargs` should span the entire circuit so that shouldn't happen.

Reply:
I don't think this is guaranteed by the current `ControlFlowOps`. e.g.
```python
qc = qk.QuantumCircuit(3,1)
with qc.while_loop((qc.clbits[0], 0)):
    qc.h(0)
    
print(qc.data[0].qubits)

# (Qubit(QuantumRegister(3, 'q'), 0),)
```

Reply:
Hmm I thought I had put that in but I guess not. I'll see about guaranteeing it.

Reply:
After talking with @jakelishman it was thought better to handle this during the routing pass. This was added in 
e51e45b4c. Not addressed yet is reducing the instruction if it has idle wires after routing.